### ISSUE TITLE
how to save images in Hive db

### ISSUE THREAD

**Issue by MyoLinOo** [hivedb/hive#546]
**Question**
Please explain the problem you are running into.

**Code sample**
```dart
Provide a few simple lines of code to show your problem.
```

**Version**
 - Platform: iOS, Android, Mac, Windows, Linux, Web
 - Flutter version: [e.g. 1.5.4]
 - Hive version: [e.g. 0.5.0]

**Comment by TheMisir** [comment]
Don't. Save images as files in disk and store its name into database.

**Comment by CodeDoctorDE** [comment]
I have a question: If I want to add offline support and web support, can I save it as bytes?

**Comment by TheMisir** [comment]
Yes you can. But I wouldn't suggest saving large binary data in database.

**Comment by X-Wei** [comment]
From the [faq](https://docs.hivedb.dev/#/more/faq?id=can-i-use-hive-to-store-binary-data-like-images):

>Can I use Hive to store binary data like images?
Yes, but don't store too big images (a few MB should be fine).

Here "a few MB" means the size of a single image or it's the size of all images?

**Comment by ricardogobbo** [comment]
I think you can stringify your image using base64 encoding and then store it in a String field.

**Comment by X-Wei** [comment]
Sure, I can also store the image blobs directly.

> I think you can stringify your image using base64 encoding and then store it in a String field. Of course, it is not the perfect way to do so, but it works fine if images won't have big sizes.

My question is rather: is it **recommended** to store e.g. >100 images (each ~1MB) in hiveDB? Will the performance be hindered?

**Comment by themisir** [comment]
> My question is rather: is it **recommended** to store e.g. >100 images (each ~1MB) in hiveDB? Will the performance be hindered?

It's really hard to answer that question. There's no "best way" in programming, it all boils down to your use cases. Hive itself is a library that simply serializes, deserializes data into a format and manages files it created. The file format itself is actually very simple.

```c
struct hive_file {
  header header; // frame count and other stuff i don't remember currently
  frame[] frames; // frames themselves
}

struct frame {
  frame_header header; // key size, key bytes, data size, type id, is deleted frame, frame crc ...
  uint8_t[] data; // binary data
}
```

As you might see putting N blob data will just stored as flat array without any post processing like compression on the hive files. Which in theory should work fine for your use case but in reality there's some things to consider:

- When you open a box the file contents will be cached in the memory. So 100 * (~1MB + a few bytes of header) will occupy device memory. For modern devices and server environment that'll be probably fine, but for low end devices this might cause OOM because your app also consumes memory for other stuff to, and when added up the memory usage might be too much for the device to handle. You can also use Lazy mode too but it'll probably perform worse than storing files separately. But if you're gonna read those files anyways into the memory, the same amount of memory will be occupied anyways, so you can use hive. In conclusion developing client apps is hard because you can't control which device are you targeting to.
- Also compaction might cause some issues depending on data size if you're doing lots of sequent updates, but you can configure it using initialization parameters when needed to find a sweet spot for your use case. (compaction is basically like defragmentation tool on OS)

**Comment by X-Wei** [comment]
Thanks for the very well-written explanation !
